nixos/mptcp: multipath TCP module

teto opened the pull request:
Multipath TCP is an IETF extension of TCP that makes use of several TCP connections to speed up transfer/increase reliability/confidentiality. Applications don't have to be changed (yet they can use the new MPTCP API for even better performance), the kernel will use MPTCP when applicable and fall back to TCP otherwise.

I had this in my repo for a while but https://github.com/NixOS/nixpkgs/issues/59301 convinced me to propose it. It only works with networkmanager for now (you need to update routing tables depending on added/removed interfaces).


<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

memberbetty reviewed (changes requested):
See http://mywiki.wooledge.org/BashFAQ/082

memberbetty commented on nixos/modules/services/networking/mptcp/mptcp_up_raw:
```diff
@@ -0,0 +1,46 @@
+fi
+
+
+if [ `grep "$DEVICE_IFACE" "$RT_TABLE" | wc -l` -eq 0 ]; then
```
`grep -c`?

memberbetty commented on nixos/modules/services/networking/mptcp/mptcp_up_raw:
```diff
@@ -0,0 +1,46 @@
+fi
+
+if [ "$DHCP4_IP_ADDRESS" ]; then
+	SUBNET=`echo $IP4_ADDRESS_0 | cut -d \   -f 1 | cut -d / -f 2`
```
Extract this code to a function with a meaningful name, because it's duplicated below.

memberbetty commented on nixos/modules/services/networking/mptcp/mptcp_up_raw:
```diff
@@ -0,0 +1,46 @@
+	echo "$NUM  $DEVICE_IFACE" >> "$RT_TABLE"
+fi
+
+if [ "$DHCP4_IP_ADDRESS" ]; then
```
Did you forget an operator here?

memberbetty commented on nixos/modules/services/networking/mptcp/default.nix:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    (mkIf (!config.networking.networkmanager.enable) {
+      warnings = [ "You have `networkmanager` disabled. Expect things to break." ];
```
Spelling (NetworkManager). Why is there only support for NetworkManager?

teto replied:
because that's the only thing I use.

memberbetty replied:
I don't use NetworkManager, but I might switch because of this change. 

Is there a reason to not enable this by default for everyone?

memberbetty commented on nixos/modules/services/networking/mptcp/mptcp_up_raw:
```diff
@@ -0,0 +1,46 @@
+
+RT_TABLE=/etc/iproute2/rt_tables
+
+if [ "$DEVICE_IFACE" = "lo" ]; then
```
Redundant quotes around lo.

teto replied:
Thanks for the comments, I can definitely improve the hook (copied from the mptcp wiki) and will do but it should already allow you to get started with mptcp (which is why I hurried with the PR :p ).

memberbetty reviewed (commented):
Interesting PR. I made some minor comments.

Mic92 commented on nixos/modules/services/networking/mptcp/mptcp_up_raw:
```diff
@@ -0,0 +1,46 @@
+
+# TODO remove
+echo $PATH
+env > /tmp/if_up_env
```
Insecure use of temporary files. `/tmp/if_up_env` could be an attacker controlled symlink.

memberbetty replied:
He said in the comment that he is going to remove that.

Mic92 replied:
I still added the comment so it won't be forgotten.

Mic92 commented on nixos/modules/services/networking/mptcp/mptcp_up_raw:
```diff
@@ -0,0 +1,46 @@
+if [ `grep "$DEVICE_IFACE" "$RT_TABLE" | wc -l` -eq 0 ]; then
+	logger -p user.info -t mptcp "Adding new routing table $DEVICE_IFACE"
+	NUM=$(wc -l < "$RT_TABLE")
+	echo "$NUM  $DEVICE_IFACE" >> "$RT_TABLE"
```
Why do we need this to be created at runtime rather then controlling the whole `etc/iproute2/rt_tables` file?

Mic92 replied:
By default `/etc/iproute2/rt_tables` does not exists, which would make:

```
NUM=$(wc -l < /etc/iproute2/rt_tables)
zsh: no such file or directory: /etc/iproute2/rt_tables
```

fail

Mic92 replied:
Is it needed at all to maintain `rt_table` for this to work?

Mic92 replied:
might be fixed by `networking.iproute2.enable = true;` so.

Mic92 commented on nixos/modules/services/networking/mptcp/mptcp_up_raw:
```diff
@@ -0,0 +1,46 @@
+	ip rule add from "$DHCP4_IP_ADDRESS" table "$DEVICE_IFACE"
+else
+	# PPP-interface
+	IPADDR=`echo $IP4_ADDRESS_0 | cut -d \   -f 1 | cut -d / -f 1`
```
ipv6?

teto replied:
what's IPv6 :D ? I haven't tested with IPv6 (haven't done much IPv6 in general). I can try but if someone wanna step up before or after the PR, I would be really happy.

Mic92 commented on nixos/modules/services/networking/mptcp/mptcp_up_raw:
```diff
@@ -0,0 +1,46 @@
+fi
+
+if [ -z "$DEVICE_IFACE" ]; then
+	logger -p user.warn -t mptcp "invalid $DEVICE_IFACE"
```
Is logging to stdout/stderr not better so it would end up in the journald log of network manager?

teto replied:
logger also ends up in journald. I used it because it was in the base example and seemed good enough: like you can set debug levels and a keyword to grep to ("mptcp" here). Now that I am looking to upstream, stdout/stderr may be better.

teto committed:
tests: add a way to retrieve test driver

Upon test failure, I sometimes want to access the VM to check
artifacts/rerun tests.

This allows to get path to the test driver via:
nixosTests.mptcp.driver.outPath

teto committed:
tests: make VM mac address deterministic

teto committed:
tests: fix interface numbering

... when qemu.networkingOptions is overriden and thus there is no
default SLIRP interface.

teto committed:
nixos/mptcp: multipath TCP module

Sane defaults for configuration of MPTCP kernel.

teto committed:
tests/mptcp: test for the multipath TCP module

Starts a multihomed VM, configure the redundant scheduler with the msh
path manager so that all paths are used.
We then launch an iperf session.

nix-build -A nixosTests.mptcp .

teto commented:
I've spent a fair deal of time looking into the nixos testing infrastructure (also to reuse it in https://github.com/Mic92/nixos-shell) and finally cleaned up the PR.
Looking forward to getting this merged: Multipath TCP is not easy to configure properly and this is really an area where nixos shines.

teto commented:
@grahamc I would very much like to reproduce the error locally. Looking at ofborg's repo, I ran this script
```
#!/bin/sh
# --argstr system thesystem 
HOME=/homeless-shelter NIX_PATH=nixpkgs=$(pwd) nix-instantiate ./nixos/release.nix -A manual --option restrict-eval true --option build-timeout 1800 --show-trace
```
to no avail (it succeeds).

teto commented:
let me alone bot, I am still on it :)

teto commented:
I am moving this module and other MPTCP-related software to https://github.com/teto/mptcp-flake.
